src/material_data.hpp

```diff
@@ -227,6 +286,35 @@ struct material_data {
     grid_size.z = 0;
     material_grid_kinds = U_DEFAULT;
   }
+
+  void copy_from(const material_data& from) {
+    which_subclass = from.which_subclass;
+    medium.copy_from(from.medium);
+
+    user_func = from.user_func;
+    // NOTE: the user_data field here opaque/void - so this is the best we can do.
+    user_data = from.user_data;
+    do_averaging = from.do_averaging;
+
+    memcpy(epsilon_dims, from.epsilon_dims, 3 * sizeof(size_t));
+    if (from.epsilon_data) {
+      int N = from.epsilon_dims[0] * from.epsilon_dims[1] * from.epsilon_dims[2];
+      epsilon_data = new double[N];
+      memcpy(epsilon_data, from.epsilon_data, N * sizeof(double));
+    }
+
+    grid_size = from.grid_size;
+    if (from.weights) {
+      int N = from.grid_size.x * from.grid_size.y * from.grid_size.z;
```

Comment:
Should this be `size_t`

Reply:
Probably this should just call `update_weights` and we should fix it there.

Reply:
Separately, we should probably add a check that the user `grid_size` makes sense (e.g. is not negative).

Reply:
Done. Made it into size_t